tests/virt_autotest/virt_autotest_base.pm

```diff
@@ -85,29 +85,88 @@ sub generateXML {
+sub post_execute_script_configuration {
+    diag("You need to override this function post_execute_script_config in your test module");
+}
+
+#This is the subroutine called inside post_execute_script_run. This is introduced to faciliate tests that do not have or not convenient
+#to use assert pattern,  can not use directly returned output from execute_script_run or needs more customized way to manipuluate results.
+#So individual test modules can also have more control over how their test results should be extracted out, intead of just relying on a
+#singel assert pattern. post_execute_script_assertion needs to be overriden in individual test modules.
+sub post_execute_script_assertion {
+    diag("You need to override this function post_execute_script_assertion in your test module");
+}
+
+#This subroutine incorporates operations needs to be done after finishing executing script. It is further dividied into three parts,
+#including post_execute_script_configuration, upload_virt_logs and do test assertion. Test assertion can be done in two ways by using
+#assert pattern or newly introduced post_execute_script_assertion subroutine.
+sub post_execute_script_run {
+    my $self = shift;
+
+    $self->post_execute_script_configuration;
+
+    if ($self->{upload_virt_log_flag} eq "yes") {
+        upload_virt_logs($self->{log_dir}, $self->{compressed_log_name});
+    }
+    save_screenshot;
+
+    my $output = $self->{script_output};
+    if ($self->{assert_pattern}) {
+        unless ($output =~ /$self->{assert_pattern}/m) {
+            assert_script_run("grep -E \"$self->{assert_pattern}\" $self->{log_dir} -r || zcat /tmp/$self->{compressed_log_name}.tar.gz | grep -aE \"$self->{assert_pattern}\"");
+        }
+    }
+    else {
+        $self->post_execute_script_assertion;
+    }
+    save_screenshot;
+}
+
+#Adding junit log and uploading guest assets are wrapped up here in newly introduced subroutine post_run_test.
+#This is called after post_execute_script_run in run_test if test passes or in post_fail_hook if test fails.
+sub post_run_test {
+    my ($self, $script_output) = @_;
+
+    if ($self->{add_junit_log_flag} eq "yes") {
+        $self->add_junit_log($script_output);
+    }
+
+    if ($self->{upload_guest_assets_flag} eq "yes") {
+        record_info('Check UPLOAD_GUEST_ASSETS flag', 'This test should upload guest assets!');
+        $self->upload_guest_assets;
+    }
+}
+
 sub execute_script_run {
     my ($self, $cmd, $timeout) = @_;
     my $pattern = "CMD_FINISHED-" . int(rand(999999));
-
     if (!$timeout) {
         $timeout = 10;
     }
 
     type_string "(" . $cmd . "; echo $pattern) 2>&1 | tee -a /dev/$serialdev\n";
-    my $ret = wait_serial($pattern, $timeout);
-
+    $self->{script_output} = wait_serial($pattern, $timeout);
     save_screenshot;
 
-    if ($ret) {
-        save_screenshot;
-        $ret =~ s/[\r\n]+$pattern[\r\n]+//g;
```

Comment:
This removal of $pattern("CMD_FINISHED-" . int(rand(999999)) shall be played on $self->{script_output}, to exclude non-real cmd output.

Reply:
Done